runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ControlClientPool.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.fnexecution.control;
+
+/** Control client pool that exposes a source and sink of control clients. */
+public interface ControlClientPool<T extends InstructionRequestHandler> {
+
+  /** Source of control clients. */
+  ClientSource<T> getSource();
+
+  /** Sink for control clients. */
+  ClientSink<T> getSink();
+
+  /** A source of control clients. */
+  @FunctionalInterface
+  interface ClientSource<T> {
+      T take() throws Exception;
+  }
+
+  /** A sink for control clients. */
+  @FunctionalInterface
+  interface ClientSink<T> {
```

Comment:
I'm not super bothered by just using `ThrowingSupplier<InstructionRequestHandler>` and `ThrowingConsumer<InstructionRequestHandler>` for these methods, instead of genericising this class. InstructionRequestHandler should in general be the only part of the interface we require.

Reply:
I originally tried that but one of those classes was in a package not accessible here. Where should those live, assuming they should be moved?

Reply:
Just checked. The problem is `ThrowingConsumer`, which lives under the `sdks/java/harness`. Is it safe to add a dependency there or should the class be moved elsewhere? This was actually the motivation for [BEAM-3966](https://issues.apache.org/jira/browse/BEAM-3966).

Reply:
In the intermediate time, move it into `sdks/java/fn-execution`